lsbatch/man1/bmgroup.1
.ds ]W %
.ds ]L
.nh
.TH bmgroup 1 "OpenLava Version 3.3 - Mar 2016"
.br
.SH NAME
\fBbmgroup\fR - displays information about host groups
.SH SYNOPSIS
.BR
.PP
.PP
\fBbmgroup\fR [\fB-r\fR] [\fB-s\fR] [\fIhost_group \fR...]
.PP
\fBbmgroup \fR[\fB-h\fR | \fB-V\fR] 
.SH DESCRIPTION
.BR
.PP
.PP
\fB\fRDisplays host groups and host names for each group.
.PP
By default, displays information about all host groups. 
.SH OPTIONS
.BR
.PP
.TP 
\fB-s
\fR
.IP
Displays GROUP_SLOT and MAX_SLOTS in addition to the host group names and group members.
Refer to lsb.hosts(5) for more information about GROUP_SLOT and MAX_SLOTS.
.IP
.TP
\fB-r\fR
.IP
Expands host groups recursively. The expanded list contains only host names.
It does not contain the names of subgroups. Duplicate names are listed only once.
.IP

.TP 
\fIhost_group \fR...

.IP
Only displays information about the specified host groups\fI. \fRDo not use 
quotes when specifying multiple host groups.


.TP 
\fB-h
\fR
.IP
Prints command usage to stderr and exits. 


.TP 
\fB-V
\fR
.IP
Prints openlava release version to stderr and exits. 


.SH OUTPUT
.BR
.PP
.PP
In the list of hosts, a name followed by a slash (/) indicates a subgroup.
.SH FILES
.BR
.PP
.PP
Host groups are defined in the configuration file 
lsb.hosts(5). 
.SH SEE ALSO
.BR
.PP
.PP
lsb.hosts(5), bugroup(1), bhosts(1)


NEWS
Release notes for OpenLava versions

4.0.0
======

-- Fix res/nios hang blocking ol make.
-- Create a new parameter NIOS_MAX_TASKS specifying the maximum number
   of tasks nios can handle.
-- Fix/reimplement btop/bbot. Since the job queue managment has undergone
   performance improvements these function had to be coded as well.
-- Fix the mbatchd variable MBD_NO_QSORT_JOBS, which was incorrectly
   set to MBD_QSORT_JOBS in daemons.c
-- Fix formatting bug in bparams.
-- Add badmin hclose/hopen -C "message", and fix parsing error of "badmin
   subcommand <command_options>".
-- Fix backward compatibility bug in which earlier version of OL cannot
   get information from higher version if mbatchd, specifically 3.0/3.1
   bqueues cannot get information from 3.3 bqueues.

3.3.3
======

-- Implement bsub -Jd to support specifying job description to jobs.
-- Implement tech preview of lsb.resources.
-- Implement the bresource command and the lsb_getlimits() API.
-- Fix scheduler bug in which the priority tree was not sorted
   after each job was dispatched.
-- Do not perform bulk slot allocation in the fairshare scheduler
   but distribute slots from higher to low priority jobs.
-- Implement periodic decay of used slots in the fairshare tree.
-- Implement a parameter HIST_MINUTES in lsb.params to control
   the time at which the decay of used slots is performed.
-- Modify the bparams command to display the HIST_MINUTES in its
   long output format.
-- The fairshare layer will log detail informations if the logclass
   in lsf.conf is either define as LC_FAIRSHARE or LC_FAIR.
-- Fix the Makefile.am to install bgmod.1 and bjdep.1 man pages.
-- Fix two memory leak in bsub -Jd implementation.
-- Introduce the lsb.resources file.
-- Implement the limit of the number of jobs per project using the
   lsb.resources file.
-- NIOS_RWAIT_SELECT must be set to y or Y in order for nios to wait
   on select to detect that a task has finished.
-- Add template lsb.resources.in.
-- Add man5 lsb.resources.5.
-- Add man1 bresources.1.
-- Implement xdrsize_ResLimitInfoReply() to compute xdr size.
-- Remove typedef of resLimitConf_t, resLimit_t, limitRes_t,
   limitConsumer_t as types are not atomic.
-- Change style declaration, e.g., char* x should be char *x.
-- Replace fname[] with gcc macro __func__.
-- Support specifying job description to job arrays with bsub -Jd.
-- Fix lsb.resources to support Unix user names that are not configured in lsb.users.
-- Implement bresources command to display current resource usage.
-- Implement project account recovery after mbd restart.
-- Add BATCH_RESLIMIT_INFO into forkOnRequest() to fork a child for this operation.
-- Fix and XDR decode error in the 3.3 mbatchd if the client has lower versions
   3.0 and 3.1.
-- Fix lsb.resources error when no user is configured in lsb.users.
-- Fix the host group expanding bug in lsb.resources.
-- Fix the resource usage counting bug when enforcing limits.
-- Release 3.3.3 the lsid command returns now 3.3.3 as OpenLava version.

3.3.2
======

-- Remove the check if a group name is also a hostname as
   there can be a host on the network not part of the cluster
   with the same name.
-- Add the noheader option for bjobs.
-- Add LSB_MAX_NUM_PROCESSORS, LSB_DJOB_NUMPROC, LSB_JOB_EXECUSER
   and LSB_EFFECTIVE_RSRCREQ variables in the job environment.
-- When a multiprocessor job (-n) specifies memory rusage and span[hosts=1]
   the memory rusage applies to the entire job instead of each slots
   of the job.
-- Fix a bug in lsb.log which was causing 'bsub -K' event being
   badly interpreted by library which printed "Bad event format".
-- Implement built in resource maxslots which represents the MBD
   MXJ parameter. This allows users to sort hosts based on MXJ.
-- Implement NUMA awareness and binding of processes based on the
   machine's NUMA architecture.
-- Fix xdr error in mbatchd if there are no job groups defined
   and user run bjgroup command.
-- Implemented the job group limit 'bgadd -L n /x/y/z'. This allows
   the users to set running job limits at different level of the job group
   hierachy. If the limit is reached jobs attached to the group will be
   pending with reason "The job group limit has been reached"
-- Renamed the bjgroups command in bjgroup.
-- Implemented bgmod to change the limits of a job group.
-- Fix mbatchd core dump when running egroup.
-- Support bsub with multiple -R options.
-- OpenLava service script sources the custom ulimit scripts.
-- Add preemption for multi-thread jobs when a host is not full.
-- Increase the daemons' port numbers in lsf.conf.in to prevent conflicts
   with other programs.
-- Increase the length of file path to PATH_MAX for bsub -i|-is|-o|-e

3.3.1
======

-- Update the lsb.queues man page describing the OWNERSHIP policy.
-- Improve error message should sbatchd failed to get go ahead
   from mbatchd,
-- Fix a bug in LIM name resolution should the caller be using
   127.0.0.1
-- Add * pattern in job name.
-- Fix bug in parsing all hosts in groups in lsb.hosts.
-- Introduced new performance enhancement parameters in lsb.params.
-- Fix res/nios blocking problems preventing olmake to run
   properly.
-- Fix nios when NIOS_RWAIT_SELECT is set in lsf.conf.

3.3.0 official release
=======

3.3.pre.2
=========
-- Keep the lsb.events and lsb.acct files always open to improve
   file access performance.
-- Implement the switching of the lsb.events asynchronously in mbatchd
   child process.
-- Implemented the bjdep command. bjdep jobid prints the status of
   job dependencies for the given job.
-- Add the bjdep man page.
-- Implement NIOS_RWAIT_SELECT, if configure in lsf.conf then nios
   and ls_rwait() will not use the SIGUSR1 to detect a task that
   has finished. ls_rwait() will select() on a connection with nios
   instead and nios will send a message when a task exits.
-- Fix RES NIOS protocol bug that prevented multiple tasks running on

lsbatch/cmd/bmgroup.c
/*
 * Copyright (C) 2015-2016 David Bigagli
 * Copyright (C) 2007 Platform Computing Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
 *
 */

#include "cmd.h"

static void prtGroups(struct groupInfoEnt *, int, int);
static void print_groups(struct groupInfoEnt *, int);

void
usage(const char *cmd)
{
    if (strstr(cmd, "bugroup") != NULL)
	fprintf(stderr, "%s [-h] [-V] [-r] [group_name ...]\n", cmd);
    else
        fprintf(stderr, "%s [-h] [-V] [-r] [-s] [group_name ...]\n", cmd);
}

int
main(int argc, char **argv)
{
    int cc;
    int numGroups = 0;
    int enumGrp = 0;
    char **groups=NULL;
    char **groupPoint;
    struct groupInfoEnt *grpInfo = NULL;
    int options = GRP_ALL;
    int all;
    int slots;

    if (lsb_init(argv[0]) < 0) {
	lsb_perror("lsb_init");
	return -1;
    }

    slots = 0;
    while ((cc = getopt(argc, argv, "Vhrs")) != EOF) {
        switch (cc) {
	    case 'r':
		options |= GRP_RECURSIVE;
		break;
	    case 'V':
		fputs(_LS_VERSION_, stderr);
		return -1;
		break;
	    case 's':
		++slots;
		break;
	    case 'h':
	    default:
		usage(argv[0]);
	    return -1;
        }
    }

    if (slots) {
	if (strstr(argv[0], "bugroup")) {
	    fprintf(stderr, "bugroup: -s option applies only to bmgroup\n");
	    return -1;
	}
    }

    numGroups = getNames(argc, argv, optind, &groups, &all, "group");
    enumGrp = numGroups;

    if (numGroups) {
        options &= ~GRP_ALL;
        groupPoint = groups;
    } else
        groupPoint = NULL;

    if (strstr(argv[0], "bugroup") != NULL) {
        options |= USER_GRP;
        grpInfo = lsb_usergrpinfo(groupPoint, &enumGrp, options);
    } else if (strstr(argv[0], "bmgroup") != NULL) {
	options |= HOST_GRP;
        grpInfo = lsb_hostgrpinfo(groupPoint, &enumGrp, options);
    }

    if (grpInfo == NULL) {
        if (lsberrno == LSBE_NO_USER_GROUP || lsberrno == LSBE_NO_HOST_GROUP ) {
            if (options & HOST_GRP)
                lsb_perror("host group");
            else
                lsb_perror("user group");
	    FREEUP(groups);
            return -1;
        }
        if (lsberrno == LSBE_BAD_GROUP && groups)

            lsb_perror (groups[enumGrp]);
        else
	    lsb_perror(NULL);
        FREEUP (groups);
	return -1;
    }

    if (numGroups != enumGrp && numGroups != 0 && lsberrno == LSBE_BAD_GROUP) {
	if (groups)
            lsb_perror (groups[enumGrp]);
        else
            lsb_perror(NULL);
        FREEUP(groups);
	return -1;
    }

    FREEUP(groups);
    if (slots == 0)
	prtGroups(grpInfo, enumGrp, options);
    else
	print_groups(grpInfo, enumGrp);

    return 0;
}

static void
prtGroups (struct groupInfoEnt *grpInfo, int numReply, int options)
{
    int i;
    int j;
    int  strLen;
    char *sp;
    char *cp ;
    char *save_sp;
    char *save_sp1;
    char word[256];
    char gname[256];
    char first = true;

    for (i = 0; i < numReply; i++) {

	sp = grpInfo[i].memberList;
	strcpy(gname, grpInfo[i].group);
	sprintf(word,"%-12.12s", gname);

	/* strip the end blank space
	 */
	strLen = strlen(gname);
	for (j = 0; j < strLen ; j++)
	    if (gname[j] == ' ') {
		gname[j] = '\0';
		break;
	    }
	if (first) {
	    if (options & USER_GRP)
		printf("%-12.12s  %s\n", "GROUP_NAME", "USERS");
	    else
		printf("%-12.12s  %s\n", "GROUP_NAME", "HOSTS");
	    first = FALSE;
	}
	printf("%s ", word);              /* print the group name */

	if (strcmp (sp, "all") == 0) {
	    if(options & USER_GRP) {
		printf("all users \n");
		continue;
	    }
	    else
		printf("all hosts used by the batch system\n");
	    continue;
	}

	/* Print out the group name.
	 */
	while ((cp = getNextWord_(&sp)) != NULL) {
	    save_sp = sp;
	    strcpy(word, cp);
	    printf("%s ", word);
	    save_sp1 = sp;
	    while ((cp = getNextWord_(&sp)) != NULL) {
		if (strcmp(word, cp) == 0) {
		    save_sp1 = strchr(save_sp1, *cp);
		    for (j = 0; j < strlen(cp); j++)
			*(save_sp1 + j) = ' ';
		}
		save_sp1 = sp;
	    }
	    sp = save_sp;
	}

	putchar('\n');
    }
}

/* print_groups()
 */
static void
print_groups(struct groupInfoEnt *groups, int num)
{
    int i;

    for (i = 0; i < num; i++) {
	printf("%-12s %s\n", "GROUP_NAME", groups[i].group);
	printf("%-12s %s\n", "HOSTS", groups[i].memberList);
	if (groups[i].group_slots[0] != 0) {
	    printf("%-12s %s\n", "GROUP_SLOT", groups[i].group_slots);
	    printf("%-12s %d\n", "MAX_SLOTS", groups[i].max_slots);
	}
    }
}
